### PR TITLE
feat(ESSNTL-4055): Add updated_start and updated_end filters

### PR DESCRIPTION
**Pull request by kruai**
# Overview

This PR is being created to address [ESSNTL-4055](https://issues.redhat.com/browse/ESSNTL-4055).
This allows users to filter the list of hosts by the `updated_date`. By providing the filters `updated_start` and `updated_end`, the user can get only the hosts that were last updated within a certain time range. I've also added a check that disallows the user from providing an `updated_start` that's later than the value for `updated_end`, if they're both provided.

In addition to adding these filters to the GET /hosts endpoint, I also added them to the DELETE /hosts and GET /tags endpoints, as they both use the common `query_filters` function to which I added these filters. I'd like to add them to more endpoints in the future, but I think we first need to refactor our filtering to reduce all the duplication ([ESSNTL-4253](https://issues.redhat.com/browse/ESSNTL-4253)).

The filter was added to xjoin with [PR 180](https://github.com/RedHatInsights/xjoin-search/pull/180).

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [x] Tests: validate optimal/expected output
- [x] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [x] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

File: tests/test_xjoin.py [hunk 1]
```diff
@@ -1742,6 +1742,50 @@ def test_xjoin_search_query_using_hostfilter_display_name(
     )
 
 
+def test_query_hosts_filter_updated_start_end(mocker, subtests, api_get, graphql_query_empty_response):
+    query_params = (
```
Comment (CodeHeeler) [inline tests/test_xjoin.py:1746]: Should we test for improperly formated date time strings as well?  Or since we're going through xjoin and have closed off the api to users directly, we aren't concerned here with improper user inputs?

File: api/filtering/filtering.py [hunk 2]
```diff
@@ -301,6 +304,17 @@ def build_registered_with_filter(registered_with):
     return ({"OR": prs_list},)
 
 
+def _build_modified_on_filter(updated_start: str = None, updated_end: str = None) -> Tuple:
+    if updated_start and updated_end and parser.isoparse(updated_start) >= parser.isoparse(updated_end):
```
[the remark this answers is not in the document]
Comment (kruai) [reply api/filtering/filtering.py:308]: isoparse won't throw a ValueError here because we explicitly call that in our validation for `format: date-time` ([validator here](https://github.com/RedHatInsights/insights-host-inventory/blob/updated-on-filter/app/validators.py#L55)). It's true that it isn't strict with the format, though, and doesn't necessarily need the T and Z markers in order to be valid; I'll test some in xjoin to see if it requires those and get back to you.
If it _does_ strictly require that format, then I'll just have to do some transformation in the filter creation.

File: tests/test_xjoin.py (hunk 1 above)
Comment (kruai) [reply tests/test_xjoin.py:1746]: We already test date validation for the swagger spec, so I don't think it's necessary to test these filters with dates like "2019-04-01TINVALID" or anything like that. However, if xjoin is strict about the format like I mentioned in my above comment, I do think I should add tests that make sure the date-time is transformed as needed.

File: api/filtering/filtering.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (kruai) [reply api/filtering/filtering.py:308]: OK, it looks like xjoin doesn't _complain_ if you send it a simplified date, but the filter doesn't work as intended. I'll add some code to transform the date-time input to the correct format and add a test for it

File: tests/test_xjoin.py (hunk 1 above)
Comment (kruai) [reply tests/test_xjoin.py:1746]: OK, I've updated the filter implementation to explicitly convert dates to a format that xjoin expects. I've also validated that xjoin works with date-time formats that don't include milliseconds or timezone info... it assumes that if your filter doesn't specify tz, you're using UTC.

File: api/filtering/filtering.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodeHeeler) [reply api/filtering/filtering.py]: Great, thanks!

**Review by CodeHeeler — APPROVED** [review]
Looks great, thanks for taking care of the date-time issues too.

**Review by strider — APPROVED** [review]
LGTM

**Review by thearifismail — APPROVED** [review]
LGTM
